src/blockstore/insert_shred.zig

```diff
@@ -0,0 +1,1975 @@
+const std = @import("std");
+const sig = @import("../lib.zig");
+
+const bs = sig.blockstore;
+const meta = bs.meta;
+const schema = bs.schema.schema;
+const shred_mod = sig.shred_collector.shred;
+const shredder = sig.blockstore.shredder;
+
+const Allocator = std.mem.Allocator;
+const ArrayList = std.ArrayList;
+const AutoHashMap = std.AutoHashMap;
+const Atomic = std.atomic.Value;
+const GetMetricError = sig.prometheus.registry.GetMetricError;
+const Mutex = std.Thread.Mutex;
+const PointerClosure = sig.utils.closure.PointerClosure;
+
+const Counter = sig.prometheus.Counter;
+const ErasureSetId = sig.shred_collector.shred.ErasureSetId;
+const Hash = sig.core.Hash;
+const Pubkey = sig.core.Pubkey;
+const Slot = sig.core.Slot;
+const Shred = sig.shred_collector.shred.Shred;
+const CodingShred = sig.shred_collector.shred.CodingShred;
+const DataShred = sig.shred_collector.shred.DataShred;
+const ReedSolomonCache = bs.shredder.ReedSolomonCache;
+const ShredId = sig.shred_collector.shred.ShredId;
+const SlotLeaderProvider = sig.core.leader_schedule.SlotLeaderProvider;
+const SortedSet = sig.utils.collections.SortedSet;
+const SortedMap = sig.utils.collections.SortedMap;
+const Timer = sig.time.Timer;
+
+const BlockstoreDB = bs.blockstore.BlockstoreDB;
+const WriteBatch = BlockstoreDB.WriteBatch;
+
+const ErasureMeta = meta.ErasureMeta;
+const Index = meta.Index;
+const MerkleRootMeta = meta.MerkleRootMeta;
+const ShredIndex = meta.ShredIndex;
+const SlotMeta = meta.SlotMeta;
+
+const serializer = sig.blockstore.database.serializer;
+
+pub const ShredInserter = struct {
+    allocator: Allocator,
+    logger: sig.trace.Logger,
+    db: BlockstoreDB,
+    lock: Mutex,
+    max_root: Atomic(u64),
+    metrics: BlockstoreInsertionMetrics,
+
+    const Self = @This();
+
+    pub fn init(
+        allocator: Allocator,
+        logger: sig.trace.Logger,
+        registry: *sig.prometheus.Registry(.{}),
```

Comment:
From the perspective of *providing* a registry, it's nice to offer a global registry. It's easy to use and helps ensure that the same registry is used everywhere.

From the perspective of *using* a registry, I think it's generally good for code to be explicit about its dependencies, rather than implicitly depending on global state. Ultimately in the normal running of a validator, it is the global registry that gets passed down, but passing it down explicitly makes it clearer what state is actually needed and used. It also makes the code more modular.